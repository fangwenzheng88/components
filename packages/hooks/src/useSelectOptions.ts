import type { SelectOptionData, SelectOptionGroup, SelectOption } from '@arco-design/web-vue'
import { isObject, isDef, isUnDef, isArray, isFunction } from '@devops-web/utils'
import { ref } from 'vue'
import type { Ref } from 'vue'

type FetchOptionsFunction = (...params: any[]) => Promise<SelectOption[]>
type SelectOptionValue = string | number | boolean | Record<string, unknown>
type FallbackFunction = (value: SelectOptionValue) => SelectOptionData

type Config = {
  valueKey: string
  fallback?: FallbackFunction
  immediate: boolean
}

/**
 * a-select 组件 options 数据的 hooks 工具
 *
 * @param fun options 数据或者一个返回 options 数据的接口函数
 * @param _config 配置
 * @param _config.valueKey 当 option 的 value 为 object 类型时，需要指定唯一标识的字段名，默认：`value`
 * @param _config.immediate 是否立即调用一次 loadOptions，默认：`false`
 * @param _config.fallback 当 value 无匹配 option 时，option 的回退函数
 *
 * @example
 * ```typescript
 * const { options } = useSelectOptions([{ label:'label-1', value:'1' }])
 *
 * // 传入函数
 * const { options, getOptions, getLabels } = useSelectOptions((params) => {
 *    console.log('传入的参数：',params)
 *    return Promise.resolve([{ label: 'label-1', value: 1 }])
 * })
 * loadOptions({type:1}) // 调用传入的函数
 *
 * // 传入函数，并立即调用loadOptions()
 * const { options, getOptions, getLabels } = useSelectOptions(() => {
 *    return Promise.resolve([{ label: 'label-1', value: 1 }])
 * },{immediate:true})
 *
 * // value为object类型
 * const { options, getOptions, getLabels } = useSelectOptions([{ label: 'label-1', value: {value:1} }])
 * getLabels(1) // label-1
 *
 * ```
 */
export function useSelectOptions(fun?: FetchOptionsFunction | SelectOption[], _config?: Partial<Config>) {
  const defaultConfig: Config = { valueKey: 'value', immediate: false }
  const config: Config = { ...defaultConfig, ..._config }

  const loading = ref(false)
  const cacheOptionsMap = new Map<string | number | boolean, SelectOptionData>()
  const options: Ref<SelectOption[]> = ref([])

  function loadOptions(...params: any[]) {
    loading.value = true
    if (isFunction(fun)) {
      return fun(...params)
        .then((data) => {
          options.value = data
          pushCacheOptions(data)
          return data
        })
        .catch((err) => {
          options.value = []
          return Promise.reject(err)
        })
        .finally(() => {
          loading.value = false
        })
    }
    return Promise.resolve([] as SelectOption[])
  }

  if (isArray(fun)) {
    options.value = fun
    pushCacheOptions(fun)
  }

  if (config.immediate) {
    loadOptions()
  }

  function isSelectOptionGroup(option: SelectOptionData | SelectOptionGroup): option is SelectOptionGroup {
    return option.isGroup === true
  }

  function pushCacheOptions(data: SelectOption[]) {
    const dataCopy = [...data]
    while (dataCopy.length > 0) {
      const option = dataCopy.pop()
      if (isDef(option)) {
        if (isObject(option)) {
          if (isSelectOptionGroup(option)) {
            dataCopy.push(...option.options)
          } else {
            const { value } = option
            if (isObject(value)) {
              cacheOptionsMap.set(value[config.valueKey] as string | number | boolean, option)
            } else if (isDef(value)) {
              cacheOptionsMap.set(value, option)
            } else {
              console.error('option：不存在value字段', option)
            }
          }
        } else {
          cacheOptionsMap.set(option, { label: String(option), value: option })
        }
      }
    }
  }

  function setOptions(data: SelectOption[]) {
    options.value = data
  }

  function getOptionFromCacheOrFallback(value: SelectOptionValue) {
    let option: SelectOptionData | undefined
    if (isObject(value)) {
      option = cacheOptionsMap.get(value[config.valueKey] as string | number | boolean)
    } else {
      option = cacheOptionsMap.get(value)
    }
    if (isUnDef(option) && isFunction(config.fallback)) {
      option = config.fallback(value)
    }
    return option
  }

  function getOptions(value: undefined | SelectOptionValue): SelectOptionData | undefined
  function getOptions(value: SelectOptionValue[]): (SelectOptionData | undefined)[]
  function getOptions(value: undefined | SelectOptionValue[]): (SelectOptionData | undefined)[] | undefined
  function getOptions(value: undefined | SelectOptionValue | SelectOptionValue[]): SelectOptionData | undefined | (SelectOptionData | undefined)[]
  function getOptions(value: undefined | SelectOptionValue | SelectOptionValue[]) {
    if (isUnDef(value)) {
      return undefined
    }
    if (Array.isArray(value)) {
      return value.map(getOptionFromCacheOrFallback)
    }
    return getOptionFromCacheOrFallback(value)
  }

  function getLabels(value: undefined | SelectOptionValue): string | undefined
  function getLabels(value: SelectOptionValue[]): (string | undefined)[]
  function getLabels(value: undefined | SelectOptionValue[]): (string | undefined)[] | undefined
  function getLabels(value: undefined | SelectOptionValue | SelectOptionValue[]): string | undefined | (string | undefined)[]
  function getLabels(value: undefined | SelectOptionValue | SelectOptionValue[]) {
    if (isUnDef(value)) {
      return undefined
    }
    const options = getOptions(value)
    if (isArray(options)) {
      return options.map((el) => {
        if (isDef(el) && isObject(el)) {
          return el.label
        }

        return el
      })
    }
    if (isUnDef(options)) {
      return undefined
    }
    if (isObject(options)) {
      return options.label
    }
    return options
  }

  function fallbackOption(value: SelectOptionValue): SelectOptionData {
    const option = getOptions(value)
    if (isDef(option)) {
      return option
    }
    if (config.fallback) {
      return config.fallback(value)
    }
    return {
      value,
      label: undefined
    }
  }
  return {
    loading,
    options,
    /**
     * 加载options，传入immediate=true，会在初始化的时候调用一次
     */
    loadOptions,
    setOptions,
    /**
     * 根据value获取对应的option，option会从之前缓存的options选项中获取，缓存的选项没有会调用传入的fallback函数并返回fallback返回的回退option，如果还没有返回undefined
     */
    getOptions,
    /**
     * 根据value获取对应的label，label会从之前缓存的options选项中获取，缓存的选项没有会调用传入的fallback函数并返回fallback返回的回退option.label，如果还没有返回undefined
     */
    getLabels,
    fallbackOption
  }
}
