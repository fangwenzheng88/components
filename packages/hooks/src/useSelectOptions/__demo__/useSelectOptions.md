# useSelectOptions

简化 `a-select` options 的定义，并在远程搜索的时候缓存历史 options，避免多选场景下远程搜素时无法展示 label 的情况

## 简单示例

:::demo

```vue
<template>
  <a-select v-model="value" class="vp-raw" style="width: 320px" placeholder="Please select ..." :options="options"></a-select>

  <div style="margin-top: 20px">
    <a-space>
      <a-button type="primary" @click="handleClick">getLabels</a-button>
    </a-space>
  </div>
</template>

<script lang="ts" setup>
import { useSelectOptions } from '@devops-web/hooks'
import { ref } from 'vue'
import { Message } from '@arco-design/web-vue'

const value = ref<string>()

const { options, getLabels } = useSelectOptions([
  { label: 'label-1', value: '1' },
  { label: 'label-2', value: '2' },
  { label: 'label-3', value: '3' }
])

function handleClick() {
  const labels = getLabels(value.value)
  Message.info(String(labels))
}
</script>
```

:::

## 远程搜索（单选）

:::demo

```vue
<template>
  <a-select
    v-model="value"
    class="vp-raw"
    style="width: 320px"
    placeholder="Please select ..."
    :options="options"
    :loading="loading"
    :allow-search="true"
    @search="loadOptions"
  ></a-select>

  <div style="margin-top: 20px">
    <a-space>
      <a-button type="primary" @click="handleClick">getLabels</a-button>
    </a-space>
  </div>
</template>

<script lang="ts" setup>
import { useSelectOptions } from '@devops-web/hooks'
import { ref } from 'vue'
import { Message } from '@arco-design/web-vue'

const value = ref<string>()

const { options, getLabels, loading, loadOptions } = useSelectOptions((inputValue: string) => {
  if (inputValue) {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve([{ label: `label-${inputValue}`, value: inputValue }])
      }, 2000)
    })
  }
  return Promise.resolve([])
})

function handleClick() {
  const labels = getLabels(value.value)
  Message.info(String(labels))
}
</script>
```

:::

## 远程搜索（多选）

:::demo

```vue
<template>
  <a-select
    v-model="value"
    class="vp-raw"
    style="width: 320px"
    placeholder="Please select ..."
    :options="options"
    :loading="loading"
    :allow-search="true"
    multiple
    :show-extra-options="false"
    @search="loadOptions"
  ></a-select>

  <div style="margin-top: 20px">
    <a-space>
      <a-button type="primary" @click="handleClick">getLabels</a-button>
    </a-space>
  </div>
</template>

<script lang="ts" setup>
import { useSelectOptions } from '@devops-web/hooks'
import { ref } from 'vue'
import { Message } from '@arco-design/web-vue'

const value = ref<string[]>([])

const { options, getLabels, loading, loadOptions } = useSelectOptions((inputValue: string) => {
  if (inputValue) {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve([{ label: `label-${inputValue}`, value: inputValue }])
      }, 2000)
    })
  }
  return Promise.resolve([])
})

function handleClick() {
  const labels = getLabels(value.value)
  Message.info(String(labels))
}
</script>
```

:::

## 远程搜索（表单回显）

:::demo

```vue
<template>
  <a-select
    v-model="value"
    class="vp-raw"
    style="width: 320px"
    placeholder="Please select ..."
    :options="options"
    :loading="loading"
    :allow-search="true"
    multiple
    @search="loadOptions"
  ></a-select>

  <div style="margin-top: 20px">
    <a-space>
      <a-button type="primary" @click="handleClick">getLabels</a-button>
    </a-space>
  </div>
</template>

<script lang="ts" setup>
import { useSelectOptions } from '@devops-web/hooks'
import { ref, reactive, onMounted } from 'vue'
import { Message } from '@arco-design/web-vue'

const details = reactive({
  selectValue: '',
  selectLabel: ''
})
const value = ref<string>()

const { options, getLabels, loading, loadOptions } = useSelectOptions(
  (inputValue: string) => {
    if (inputValue) {
      return new Promise((resolve) => {
        setTimeout(() => {
          resolve([{ label: `label-${inputValue}`, value: inputValue }])
        }, 2000)
      })
    }
    return Promise.resolve([])
  },
  {
    fallback(value) {
      if (value === details.selectValue) {
        return {
          label: details.selectLabel,
          value: details.selectValue
        }
      }
      return { value: value as string, label: undefined }
    }
  }
)

onMounted(() => {
  setTimeout(() => {
    Object.assign(details, {
      selectValue: '999',
      selectLabel: 'label-999'
    })
    value.value = details.selectValue
  }, 1000)
})

function handleClick() {
  const labels = getLabels(value.value)
  Message.info(String(labels))
}
</script>
```

:::
