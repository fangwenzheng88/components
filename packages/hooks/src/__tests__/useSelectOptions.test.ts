import { test, expect, describe } from 'vitest'
import { useSelectOptions } from '../useSelectOptions'

describe('useSelectOptions', () => {
  test('入参为普通options数组', () => {
    const { options, getOptions, getLabels } = useSelectOptions([{ label: 'label-1', value: 1 }])
    expect(options.value).toMatchSnapshot()
    expect(getOptions(1)).toEqual({ label: 'label-1', value: 1 })
    expect(getLabels(1)).toBe('label-1')
  })

  test('入参为options数组，value为对象类型', () => {
    const { options, getOptions, getLabels } = useSelectOptions([{ label: 'label-1', value: { value: 1 } }])
    expect(options.value).toMatchSnapshot()
    expect(getOptions(1)).toEqual({ label: 'label-1', value: { value: 1 } })
    expect(getOptions({ value: 1 })).toEqual({ label: 'label-1', value: { value: 1 } })
    expect(getLabels({ value: 1 })).toBe('label-1')
    expect(getLabels(1)).toBe('label-1')
  })

  test('入参为options数组，value为对象类型，自定义valueKey', () => {
    const { options, getOptions, getLabels } = useSelectOptions([{ label: 'label-1', value: { valueKey: 1 } }], { valueKey: 'valueKey' })
    expect(options.value).toMatchSnapshot()
    expect(getOptions(1)).toEqual({ label: 'label-1', value: { valueKey: 1 } })
    expect(getOptions({ valueKey: 1 })).toEqual({ label: 'label-1', value: { valueKey: 1 } })
    expect(getLabels({ valueKey: 1 })).toBe('label-1')
    expect(getLabels(1)).toBe('label-1')
  })

  test('入参为函数', () => {
    const { options, getOptions, getLabels } = useSelectOptions(
      () => {
        return Promise.resolve([{ label: 'label-1', value: 1 }])
      },
      { immediate: true }
    )
    setTimeout(() => {
      expect(options.value).toMatchSnapshot()
      expect(getOptions(1)).toEqual({ label: 'label-1', value: 1 })
      expect(getLabels(1)).toBe('label-1')
    }, 1000)
  })

  test('配置fallback', () => {
    const { options, getOptions, getLabels } = useSelectOptions([], {
      fallback(value) {
        if (value === 1) {
          return { value, label: 'label-1' }
        } else {
          return { value, label: undefined }
        }
      }
    })
    expect(getOptions(1)).toEqual({ label: 'label-1', value: 1 })
    expect(getLabels(1)).toBe('label-1')

    expect(getOptions(2)).toEqual({ label: undefined, value: 2 })
    expect(getLabels(2)).toBe(undefined)
  })
})
