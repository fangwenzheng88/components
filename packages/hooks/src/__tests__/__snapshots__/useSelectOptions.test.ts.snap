// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useSelectOptions > 入参为options数组，value为对象类型 1`] = `
[
  {
    "label": "label-1",
    "value": {
      "value": 1,
    },
  },
]
`;

exports[`useSelectOptions > 入参为options数组，value为对象类型，自定义valueKey 1`] = `
[
  {
    "label": "label-1",
    "value": {
      "valueKey": 1,
    },
  },
]
`;

exports[`useSelectOptions > 入参为函数 1`] = `[]`;

exports[`useSelectOptions > 入参为普通options数组 1`] = `
[
  {
    "label": "label-1",
    "value": 1,
  },
]
`;
